cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java

```diff
@@ -413,10 +394,58 @@ private void validate() {
 
     static class EnhancerFactoryData {
         public final Class generatedClass;
-        public volatile Factory factory;
+        private final Method setThreadCallbacks;
+        private final Class[] primaryConstructorArgTypes;
+        private final Constructor primaryConstructor;
 
-        public EnhancerFactoryData(Class generatedClass) {
+        public EnhancerFactoryData(Class generatedClass, Class[] primaryConstructorArgTypes, boolean classOnly) {
             this.generatedClass = generatedClass;
+            try {
+                setThreadCallbacks = getCallbacksSetter(generatedClass, SET_THREAD_CALLBACKS_NAME);
+                setThreadCallbacks.setAccessible(true);
+                if (classOnly) {
+                    this.primaryConstructorArgTypes = null;
+                    this.primaryConstructor = null;
+                } else {
+                    this.primaryConstructorArgTypes = primaryConstructorArgTypes;
+                    this.primaryConstructor = generatedClass.getDeclaredConstructor(primaryConstructorArgTypes);
+                    primaryConstructor.setAccessible(true);
+                }
+            } catch (NoSuchMethodException e) {
+                throw new CodeGenerationException(e);
+            }
+        }
+
+        public Object newInstance(Class[] argumentTypes, Object[] arguments, Callback[] callbacks) {
+            setThreadCallbacks(callbacks);
+            try {
+                if (!Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
+                    return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
+                }
+                try {
+                    return primaryConstructor.newInstance(arguments);
+                } catch (InstantiationException e) {
+                    throw new CodeGenerationException(e);
+                } catch (IllegalAccessException e) {
+                    throw new CodeGenerationException(e);
+                } catch (InvocationTargetException e) {
+                    throw new CodeGenerationException(e.getTargetException());
+                }
+            } finally {
+                // clear thread callbacks to allow them to be gc'd
+                setThreadCallbacks(null);
+            }
+
+        }
+
+        private void setThreadCallbacks(Callback[] callbacks) {
```

Comment:
Frankly speaking, I do not like the idea of `threadCallbacks` at all. I would rather call a constructor that accepts callbacks, and uses accordingly.

Back to the topic:
1) `Enhancer.setThreadCallbacks` does not cache `java.lang.reflect.Method`
2) `EnhancerFactoryData` caches `reflect.Method`, that is what makes this invocation fast.

`EnhancerFactoryData` is private, so I do not expect users to call it or even know of its existence.

Re `Enhancer.setThreadCallbacks`, I would suggest to add a note that the method might be slow (or wire cache there somehow).